resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/rest/kubernetes/ResourceStagingServerSuite.scala

```diff
@@ -125,6 +134,44 @@ class ResourceStagingServerSuite extends SparkFunSuite with BeforeAndAfter {
     val downloadedBytes = ByteStreams.toByteArray(responseBody.byteStream())
     assert(downloadedBytes.toSeq === bytes)
   }
+
+  private def startServer(): Int = {
+    var currentAttempt = 0
+    var successfulStart = false
+    var latestServerPort = new ServerSocket(0).getLocalPort
+    while (currentAttempt < MAX_SERVER_START_ATTEMPTS && !successfulStart) {
+      val newServer = new ResourceStagingServer(latestServerPort, serviceImpl, sslOptionsProvider)
+      try {
+        newServer.start()
+        successfulStart = true
+        server = Some(newServer)
+      } catch {
+        case e: Throwable =>
+          try {
+            newServer.stop()
+          } catch {
+            case e1: Throwable =>
+              log.warn("Failed to stop a resource staging server that failed to start.", e1)
+          }
+
+          if (Utils.isBindCollision(e)) {
+            currentAttempt += 1
+            latestServerPort = latestServerPort + 1
```

Comment:
could do a randomization strategy more effective than `+=1` but should already significantly reduce chances of collision